docs/architecture/adr-021-protobuf-query-encoding.md

```diff
@@ -0,0 +1,238 @@
+grpc plugin, which for a service named `Query` generates an interface named
+`QueryServer` as below:
+
+```go
+type QueryServer interface {
+	QueryBalance(context.Context, *QueryBalanceParams) (*types.Coin, error)
+	QueryAllBalances(context.Context, *QueryAllBalancesParams) (*QueryAllBalancesResponse, error)
+}
+```
+
+The custom queries for our module are implemented by implementing this interface.
+
+The first parameter in this generated interface is a generic `context.Context`,
+whereas querier methods generally need an instance of `sdk.Context` to read
+from the store. Since arbitrary values can be attached to `context.Context`
+using the `WithValue` and `Value` methods, the SDK should provide a function
+`sdk.UnwrapSDKContext` to retrieve the `sdk.Context` from the provided
+`context.Context`.
+
+An example implementation of `QueryBalance` for the bank module as above would
+look something like:
+
+```go
+type Querier struct {
+	Keeper
+}
+
+func (q Querier) QueryBalance(ctx context.Context, params *types.QueryBalanceParams) (*sdk.Coin, error) {
+	balance := q.GetBalance(sdk.UnwrapSDKContext(ctx), params.Address, params.Denom)
+	return &balance, nil
+}
+```
+
+### Custom Query Registration and Routing
+
+Query server implementations as above would be registered with `AppModule`s using
+a new method `RegisterQueryServer(grpc.Server)` which could be implemented simply
+as below:
+
+```go
+// x/bank/module.go
+func (am AppModule) RegisterQueryServer(server grpc.Server) {
+	types.RegisterQueryServer(server, keeper.Querier{am.keeper})
+}
+```
+
+Underneath the hood, a new method `QueryRouter.RegisterService(sd *grpc.ServiceDesc, handler interface{})`
+would be added to actually add the queries to the routing table.
+
+GRPC-like requests are routed by the service name (ex. `cosmos_sdk.x.bank.v1.Query`)
+and method name (ex. `QueryBalance`) combined with `/`s to form a full
+method name (ex. `/cosmos_sdk.x.bank.v1.Query/QueryBalance`). This gets translated
+into an ABCI query as `custom/cosmos_sdk.x.bank.v1.Query/QueryBalance`. Beyond the
+method name, GRPC requests carry a protobuf encoded payload - mapped naturally
+to `RequestQuery.Data` and receive a protobuf encoded response or error. Thus
+there is a quite natural mapping of GRPC-like rpc methods to the existing
+`sdk.Query` and `QueryRouter` infrastructure.
+
+This basic specification allows us to reuse protocol buffer `service` definitions
+for ABCI custom queries substantially reducing the need for manual decoding and
+encoding in query methods. 
+
+### GRPC Protocol SUpport
+
+In addition to providing an ABCI query pathway, we can easily provide a GRPC
+proxy server that routes requests in the GRPC protocol to ABCI query requests
+under the hood. In this way, clients could use their host languages' existing
+GRPC implementations to make direct queries against Cosmos SDK app's using
+these `service` definitions.
+
```

Comment:
What, if any, changes need to be made to BaseApp in order to facilitate the use of gRPC?